Add Dockerfile for WebKitGTK testing (based on Debian)

clopez opened the pull request:
This PR is like #15580 but it adds a dockerfile based on Debian experimental instead of fedora rawhide, because webdriver testing with webkit is currently broken in fedora. Once the fedora situation gets better, this dockerfile can get replaced with one based on fedora like the one from #15580

To test this do the following:

0. Checkout this branch
1. Build the dockerfile
```
cd tools/docker
docker build -f Dockerfile.epiphany -t wpt_debian_epiphany .
```
2. Run epiphany infrastructure tests with:
```
docker run --rm -it  wpt_debian_epiphany /bin/bash --login -c "set -ex; ~/start.sh https://github.com/clopez/wpt pr-epiphany-infrastructure-expectations; cd ~/web-platform-tests; sudo sh -c './wpt make-hosts-file >> /etc/hosts'; NO_AT_BRIDGE=1 xvfb-run -a -s '-screen 0 1024x768x24' dbus-run-session ./wpt run --metadata=infrastructure/metadata --webdriver-arg='--host=127.0.0.1' epiphany infrastructure"
```
* Note that in the previous command its fetching from my branch of wpt with the epiphany expectations for infrastructure. See PR #18589 for the merge request of this.

clopez commented:
@foolip I managed to get taskcluster running for epiphany on this branch of my fork https://github.com/clopez/wpt/commits/test-taskcluster-epiphany 
it using the docker container image of this PR.

The results of a complete run can be seen here: https://tools.taskcluster.net/groups/SasDK-a9QRiUcU7FX__n1A 

I uploaded them to staging.wpt.fyi here: https://staging.wpt.fyi/results/?run_id=252960012

foolip commented:
Very cool, thank you @clopez! The next thing I'd do is look at infrastructure/ results and to compare the results against other runs to see if there's anything suspicious.

Looks like you branched from commit dbb4e0ccfd0a97f80becd851371fb5a205aef7ec, for which we do have results to compare to: https://staging.wpt.fyi/results/?sha=dbb4e0ccfd, with https://staging.wpt.fyi/results/?run_id=256970019 (STP) being the most useful reference.

Epiphany results have only 33250 tests, compared to 34396 tests for Safari. Those numbers should be the same, modulo the tests disabled by #18585.

In infrastructure/, document-fonts-ready.html failing is like a problem.

https://staging.wpt.fyi/results/?diff&filter=ADC&run_id=256970019&run_id=252960012 compares against the STP run and there are many differences, but looking closely at that isn't worthwhile until infrastructure/ differences are understood and all results available.

clopez commented:
> Very cool, thank you @clopez! The next thing I'd do is look at infrastructure/ results and to compare the results against other runs to see if there's anything suspicious.
> 
> Looks like you branched from commit [dbb4e0c](https://github.com/web-platform-tests/wpt/commit/dbb4e0ccfd0a97f80becd851371fb5a205aef7ec), for which we do have results to compare to: https://staging.wpt.fyi/results/?sha=dbb4e0ccfd, with https://staging.wpt.fyi/results/?run_id=256970019 (STP) being the most useful reference.
> 
> Epiphany results have only 33250 tests, compared to 34396 tests for Safari. Those numbers should be the same, modulo the tests disabled by #18585.
> 
One of the taskcluster steps failed to complete for epiphany (the WPT runner hangs and taskcluster kills the instance due to timeout). That may explain this difference of results

> In infrastructure/, document-fonts-ready.html failing is like a problem.
>

That is a expected failure, it has been marked as such in the expectations for epiphany/webkitgtk I submitted in Pull-Request #18589

It is due to this WK bug https://bugs.webkit.org/show_bug.cgi?id=174030 that I think it should affect last Safari Tech Preview (90), but wpt.fyi is still testing with Safari Tech Preview 82 
 
> https://staging.wpt.fyi/results/?diff&filter=ADC&run_id=256970019&run_id=252960012 compares against the STP run and there are many differences, but looking closely at that isn't worthwhile until infrastructure/ differences are understood and all results available.

foolip commented:
Yes, document-fonts-ready.html should indeed fail on STP as well, the reason it doesn't is because we're still on macOS High Sierra with Ahem installed as a system font.

For Epiphany, did you install Ahem as a system font?

clopez commented:
> Yes, document-fonts-ready.html should indeed fail on STP as well, the reason it doesn't is because we're still on macOS High Sierra with Ahem installed as a system font.
> 
> For Epiphany, did you install Ahem as a system font?

No I didn't.
Installing Ahem as system font is something that needs to be done when creating the Docker container or is something that the script run_tc.py should do itself when starting the tests?
I don't see how this is done for the other Dockerfile with Ubuntu 18.04.

jgraham commented:
The dependency on Ahem as a system font is currently being removed, but in the meantime running with `--install-fonts` is supposed to put it into the user local font directory on Linux. If there's something with the epiphany setup that means it's not being read from there then we either need a different solution or to ignore the problem until Ahem isn't required any longer.

clopez commented:
> The dependency on Ahem as a system font is currently being removed, but in the meantime running with `--install-fonts` is supposed to put it into the user local font directory on Linux. If there's something with the epiphany setup that means it's not being read from there then we either need a different solution or to ignore the problem until Ahem isn't required any longer.

Ups, I missed that ```--install-fonts``` switch that it uses. Thanks.
Yes, it seems to be working as expected. The Ahem test is passing: https://staging.wpt.fyi/results/infrastructure/assumptions?run_id=252960012

clopez commented:
I have another question about infrastructure tests. I see that for the epiphany run there are tests failing (like document-fonts-ready.html) that in theory I already marked as expected failures for epiphany.

And I don't see the taskcluster wpt runner passing the parameter ```--metadata infrastructure/metadata``` so it gets the expectations.


Should I modify the command line arguments in ```.taskcluster.yml``` so it passes this parameter for epiphany?

clopez commented:
> Should I modify the command line arguments in `.taskcluster.yml` so it passes this parameter for epiphany?

Tried doing this but didn't helped. See run for this from https://github.com/clopez/wpt/commit/b1ace2fcfe194d8c9b28fc4759b8b28d3dde01df at:
https://tools.taskcluster.net/groups/c1l4_RRsTJyo1kvQE2-Uuw
https://staging.wpt.fyi/results/?run_id=197440003

psaavedra commented:
> > Epiphany results have only 33250 tests, compared to 34396 tests for Safari. Those numbers should be the same, modulo the tests disabled by #18585.
> 
> One of the taskcluster steps failed to complete for epiphany (the WPT runner hangs and taskcluster kills the instance due to timeout). That may explain this difference of results
> 

Finally,  I found out the reason why this hang is happening. Sometimes the previous WebDriverTest server (from the previous `run_test()`)  doesn't stop by a SIGTERM signal and the `mozprocess.kill()` triggered  by the  `TestRunnerManager::stop_runner()` method keeps waiting _ad infinitum_ in the `.wait()`. I made a quick hack to bypass this.

The following workaround works avoiding the hang forcing a `SIGKILL` instead of a `SIGTERM` for the WebDriverTest server :

* Task runner output: 
  * https://tools.taskcluster.net/task-group-inspector/#/WJrYHUsLTeiTAe9Jy6BPhw

* Commits:
  * https://github.com/psaavedra/wpt/commit/35b7d9a453dd274abef94d96d1e74723c45e9a7d
  * https://github.com/psaavedra/wpt/commit/0ba934025059a4dec29e750439fe2431e1aed8bd

foolip commented:
Regarding https://github.com/web-platform-tests/wpt/pull/18595#issuecomment-526252665, the expectations for the infrastructure/ tests (in infrastructure/metadata/infrastructure/) are used only for running the infrastructure/ tests specifically in CI, to ensure that these do not regress without us noticing.

When doing a full run of wpt, we don't use the expectations metadata. I think doing so would be harmless since wpt.fyi uses the actual results in the report without comparing to expected, but it would also be unnecessary. @lukebjerring @Hexcles is that all correct?

foolip commented:
@psaavedra the WebDriverTest issue you've identified sounds like it could be similar to what happened in https://github.com/web-platform-tests/wpt/issues/16229. @gsnedders may have ideas for what a correct workaround would be.

Is there a WebKitDriver bug at the bottom of this that should be reported as well?

foolip commented:
@psaavedra do you have an account for wpt.fyi so that you could upload the results of https://tools.taskcluster.net/task-group-inspector/#/WJrYHUsLTeiTAe9Jy6BPhw? @Hexcles can get you set up with that if you don't have one.

foolip commented:
@psaavedra, alternatively, you could work on a branch of the main wpt repo, which would cause each run to be uploaded automatically. If you'd prefer that, I can give you write access to the repo.

psaavedra commented:
> @psaavedra the WebDriverTest issue you've identified sounds like it could be similar to what happened in #16229. @gsnedders may have ideas for what a correct workaround would be.
> 
> Is there a WebKitDriver bug at the bottom of this that should be reported as well?

For our case,  I suppose we could just skip the `print-microtask-after-navigate.htm` with: `--exclude /printing/print-microtask-after-navigate.html`. Apparently this is the one who is causing the problem:

```
TEST-UNEXPECTED-ERROR | /printing/print-microtask-after-navigate.html | Callback was not called before the unload event.
Traceback (most recent call last):
  File "/home/test/web-platform-tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 294, in _run
    self.result = True, self.func(self.protocol, self.url, self.timeout)
  File "/home/test/web-platform-tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 370, in do_testharness
    self.script_resume % format_map, async=True)
  File "/home/test/web-platform-tests/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 39, in execute_script
    return method(script)
  File "/home/test/web-platform-tests/tools/webdriver/webdriver/client.py", line 20, in inner
    return func(self, *args, **kwargs)
  File "/home/test/web-platform-tests/tools/webdriver/webdriver/client.py", line 650, in execute_async_script
    return self.send_session_command("POST", "execute/async", body)
  File "/home/test/web-platform-tests/tools/webdriver/webdriver/client.py", line 516, in send_session_command
    return self.send_command(method, url, body)
  File "/home/test/web-platform-tests/tools/webdriver/webdriver/client.py", line 480, in send_command
    raise err
JavascriptErrorException: javascript error (500): Callback was not called before the unload event.
```

Regarding if this is a problem in WebKitDriver to report upstream. Not sure. Let me make some extra checks.

psaavedra commented:
> @psaavedra do you have an account for wpt.fyi so that you could upload the results of https://tools.taskcluster.net/task-group-inspector/#/WJrYHUsLTeiTAe9Jy6BPhw? @Hexcles can get you set up with that if you don't have one.

Trying to figure out how to upload the results. Apparently we have credentials for `igalia` so ... let try it.

psaavedra commented:
> > @psaavedra do you have an account for wpt.fyi so that you could upload the results of https://tools.taskcluster.net/task-group-inspector/#/WJrYHUsLTeiTAe9Jy6BPhw? @Hexcles can get you set up with that if you don't have one.
> 
> Trying to figure out how to upload the results. Apparently we have credentials for `igalia` so ... let try it.

I think I managed myself to send the report:

```
Sending a total of 26 wptresults and 26 screenshoots to https://staging.wpt.fyi
Task 253050003 added to queue

Check status in: https://staging.wpt.fyi/results/?run_id=253050003
```

... the UI still doesn't show anything but I think it is because it is queued , right?

foolip commented:
Yeah, it must have been in the queue, now the results are visible. The number of tests look right, 34438 just like in https://staging.wpt.fyi/results/?run_id=231070009.

Comparing against the same STP run as in https://github.com/web-platform-tests/wpt/pull/18595#issuecomment-526071506:
https://staging.wpt.fyi/results/?diff&filter=ADC&run_id=256970019&run_id=253050003

document-fonts-ready.html is still failing, which I wouldn't expect if you're now using `--install-fonts`. But even without `--install-fonts`, https://bugs.webkit.org/show_bug.cgi?id=174030 is now fixed, so it'd be very interesting to see if document-fonts-ready.html with a recent WebKit binary.

Beyond that, the failures in infrastructure/server/ would be good to understand. Looks like Service Workers aren't enabled, which will result in a lot of failures all over the place.

psaavedra commented:
> Service Workers



> Yeah, it must have been in the queue, now the results are visible. The number of tests look right, 34438 just like in https://staging.wpt.fyi/results/?run_id=231070009.
> 
> Comparing against the same STP run as in [#18595 (comment)](https://github.com/web-platform-tests/wpt/pull/18595#issuecomment-526071506):
> https://staging.wpt.fyi/results/?diff&filter=ADC&run_id=256970019&run_id=253050003
> 

Great.

> document-fonts-ready.html is still failing, which I wouldn't expect if you're now using `--install-fonts`. But even without `--install-fonts`, https://bugs.webkit.org/show_bug.cgi?id=174030 is now fixed, so it'd be very interesting to see if document-fonts-ready.html with a recent WebKit binary.
> 

(nod)

> Beyond that, the failures in infrastructure/server/ would be good to understand. Looks like Service Workers aren't enabled, which will result in a lot of failures all over the place.

I see @clopez set on this flag weeks ago but dunno if the image that I'm using right contains this change. 

Let's see what happens bumping up webkit in the Docker image

foolip commented:
When it comes to differences outside of infrastructure/, filtering to `seq(status:ok status:error)` in wpt.fyi shows differences that are IMHO extra worth look at. (One has to remove the filter to understand subtest results, see https://github.com/web-platform-tests/wpt.fyi/issues/1456.)

One difference that seems like it likely points to a real problem is in pointerevents/pointerevent_attributes_hoverable_pointers.html, where the error is "WebDriverException: element not interactable".

psaavedra commented:
> > Service Workers
> 
> > Yeah, it must have been in the queue, now the results are visible. The number of tests look right, 34438 just like in https://staging.wpt.fyi/results/?run_id=231070009.
> > Comparing against the same STP run as in [#18595 (comment)](https://github.com/web-platform-tests/wpt/pull/18595#issuecomment-526071506):
> > https://staging.wpt.fyi/results/?diff&filter=ADC&run_id=256970019&run_id=253050003
> 
> Great.
> 
> > document-fonts-ready.html is still failing, which I wouldn't expect if you're now using `--install-fonts`. But even without `--install-fonts`, https://bugs.webkit.org/show_bug.cgi?id=174030 is now fixed, so it'd be very interesting to see if document-fonts-ready.html with a recent WebKit binary.
> 
> (nod)
> 
> > Beyond that, the failures in infrastructure/server/ would be good to understand. Looks like Service Workers aren't enabled, which will result in a lot of failures all over the place.
> 
> I see @clopez ~set on~/ (created a bug for this in bugzilla [1]) ... this flag weeks ago but dunno if the image that I'm using right contains this change.
> 
> [1] https://bugs.webkit.org/show_bug.cgi?id=200815
> 
> Let's see what happens bumping up webkit in the Docker image

Ey @foolip , finally I managed myself to build WebKit with `Service worker` enabled. I was a bit stuck because I was not able to activate the functionality only by adding the `-DENABLE_SERVER_WORKER=ON` flag so I add to build the package with `experimental` set to ON and fix the dependencies. Having said that, I've created a custom Debian packages reusing and forking [1] the Berto García official packages for Debian.

... so I've uploaded those packages to a temporary HTTP server [2] and also tuned the Debian docker image to force the installation of those packages in the image.

A `wpt run` of this image recognises the `navigator.serviceWorker.getRegistration()` now, at least and it is passing tests (not all).

Finally, I've launched the testrunner [3]  minutes ago  for this image. I will upload the results once it  finish to compare the results with those for STP.

I think it worth to have some kind of CI building this custom .deb for unstable webkit with custom flags activated and patches. 

.... ah! I also applied [5] in this package.

[1] https://gitlab.com/saavedra.pablo/webkit-debian/commits/psaavedra/wk2/experimental
[2] https://wk-contrib.igalia.com/debian/tmp/
[3] https://github.com/psaavedra/wpt/commit/df2a078d64cb2e59fbbe87837a4831b98822a993
[4] https://tools.taskcluster.net/task-group-inspector/#/ejOc7hbbR16mAXRnl6_TdQ
[5]  https://bugs.webkit.org/show_bug.cgi?id=174030

Hexcles commented:
> When doing a full run of wpt, we don't use the expectations metadata. I think doing so would be harmless since wpt.fyi uses the actual results in the report without comparing to expected, but it would also be unnecessary. @lukebjerring @Hexcles is that all correct?

That's all correct.

psaavedra commented:
> 
> Finally, I've launched the testrunner [3] minutes ago for this image. I will upload the results once it finish to compare the results with those for STP.
> 
Here they are ...

https://staging.wpt.fyi/results/?diff&filter=ADC&q=seq%28status%3Aok%20status%3Aerror%29&run_id=256970019&run_id=257120006

looks better respect the others:

* Diffs in 2dcontents, content-security-policy, css and wasm respect STP are gone.
* fetch only has 1 difference
* webrtc is explicitly disabled because dependencies issues.
* ... and still diffs for: cookies, pointerevents, pointerlock,
* upgrade-insecure-requests, workers
* ./infrastructure/assumptions/document-fonts-ready.html: is OK now [1]

The counterpart is offscreen-canvas start to fail.

[1] https://staging.wpt.fyi/results/infrastructure/assumptions/document-fonts-ready.html?run_id=257120006

foolip commented:
> ./infrastructure/assumptions/document-fonts-ready.html: is OK now

Yay!!

The diff between STP and unstable webkit filtered to `seq(status:ok status:error)` is quite modest now. Other searches worth looking at are `status:unknown` to find missing tests or subtests, and `seq(status:pass status:!pass)` to look at more straightforward regressions. Among those are some that are because of `navigator.mediaDevices.getUserMedia` not being exposed. Accounting for most of the differences and enabling experimental features to fix it where possible would be great.

For setting up CI to build webkit regularly, how much machine time is needed? Presumably nothing special is needed and we could put this on any of Taskcluster, Azure Pipelines or GitHub Actions, in this repo or in a helper repo.

psaavedra commented:
> For setting up CI to build webkit regularly, how much machine time is needed? Presumably nothing special is needed and we could put this on any of Taskcluster, Azure Pipelines or GitHub Actions, in this repo or in a helper repo.

Well it is about to compile WebKit from the scratch the bigger machine the better. As reference:

* CPU: 2 x Xeon SP Silver 4114 (13.75M L3 Cache, 2.20 to 3.00 GHz per CPU) (20 cores / 40 threads total)
* RAM: 64GB DDR4 2400
* I/O: Samsung 960 EVO NVMe
* GCC: 7.1.1 20170622 (Red Hat 7.1.1-3) (GCC) 
* OS: Fedora 26
* Results: time ninja:

**Time to build WebKit**:         14 minutes aprox. from the scratch.

To simplify we can say time grows and comes down lineally respect the CPU cores. 

To be in the same page. AFAIK, TestRunner uses Docker to run the Runner.  The build package process relies in `gbp buildpackage` (git-buildpackage) wich uses `pbuilder`  and requires `sudo` and some `CAPS` (for example to allow `mknod`) in the runner.  is this possible with the TestRunner workers based on Docker? Alternatively, I think we could setup a bot in https://build-webkit.igalia.com.

psaavedra commented:
> > ./infrastructure/assumptions/document-fonts-ready.html: is OK now
> 
> Yay!!
> 
> The diff between STP and unstable webkit filtered to `seq(status:ok status:error)` is quite modest now. Other searches worth looking at are `status:unknown` to find missing tests or subtests, and `seq(status:pass status:!pass)` to look at more straightforward regressions. Among those are some that are because of `navigator.mediaDevices.getUserMedia` not being exposed. Accounting for most of the differences and enabling experimental features to fix it where possible would be great.
> 
> For setting up CI to build webkit regularly, how much machine time is needed? Presumably nothing special is needed and we could put this on any of Taskcluster, Azure Pipelines or GitHub Actions, in this repo or in a helper repo.



> Other searches worth looking at are `status:unknown` to find missing tests or subtests

OK. If I catched the  idea it seems here I have bunch of errors related with webdriver to work on it: https://staging.wpt.fyi/results/webdriver/tests?diff&filter=ADC&q=status%3Aunknown&run_id=256970019&run_id=257120006

foolip commented:
Yes, absolutely, differences in webdriver/ results could affect lots of other things since we depend on WebDriver to execute the tests. Understanding those differences and fixing the ones that could plausibly affect the execution of other tests would be great.

foolip commented:
As for building/archiving WebKit, it sounds like the requirements are quite modest. Is https://build-webkit.igalia.com/ part of WebKit's CI, or is it a continuous builder that doesn't show up in WebKit code review?

Are WebKit LayoutTests (which include WPT) run on Linux as part of WebKit CI? If so, what produces those builds? We can't use the results of those runs directly because (like Chromium) `./wpt run` isn't used and no JSON report is produced, but I wonder how much configuration and infrastructure can be shared? In particular, if some settings are enabled to make WPT pass on Linux for WebKit devs, using those same settings for full wpt.fyi runs would be ideal.

Ms2ger commented:
https://build.webkit.org/builders/GTK%20Linux%2064-bit%20Release%20%28Tests%29 is a test builder that run the in-tree wpt tests; I believe it gets the build from https://build.webkit.org/builders/GTK%20Linux%2064-bit%20Release%20Debian%20Stable%20%28Build%29 with `./Tools/Scripts/build-webkit`.

psaavedra commented:
> As for building/archiving WebKit, it sounds like the requirements are quite modest. Is 

even the CAPABILITIES thing?. I am newbie with taskcluster. how it suppose how I could integrate this .deb build task on it?

> https://build-webkit.igalia.com/ part of WebKit's CI, or is it a continuous builder that doesn't show up in WebKit code review?
> 
Actually this is a fork of the real build.webkit.org. Usually, we use this as staying to test new builders or steps before land it into the official.

> Are WebKit LayoutTests (which include WPT) run on Linux as part of WebKit CI? If so, what produces those builds? We can't use the results of those runs directly because (like Chromium) `./wpt run` isn't used and no JSON report is produced, but I wonder how much configuration and infrastructure can be shared? In particular, if some settings are enabled to make WPT pass on Linux for WebKit devs, using those same settings for full wpt.fyi runs would be ideal.

Not in the Layout test CI  but, for example, there is a worker running WPT and uploading the results for MathML: https://build-webkit.igalia.com/builders/GTK%20Linux%2064-bit%20Release%20%28WPT%20Tests%29/builds/1068

foolip commented:
> even the CAPABILITIES thing?

That I'm not sure about, I just mean it doesn't require very much compute time, so should be cheap to run *somewhere*.

> Not in the Layout test CI

Are LayoutTests run only on macOS then for every WebKit change?

foolip commented:
@Ms2ger getting a compatible report out of WPT-in-LayoutTests isn't straightforward since no results at all are recorded anywhere if everything is passing. At least that's the case for Chromium. One could of course tweak the LayoutTests runner to make this work, but using wptrunner avoid this problem, and also makes it possible to test the exact same revision of WPT in multiple browsers, which helps diffing of course.

psaavedra commented:
> > even the CAPABILITIES thing?
> 
> That I'm not sure about, I just mean it doesn't require very much compute time, so should be cheap to run _somewhere_.

Ok let give me a try it in my Docker instance to see if we can follow this path.

> 
> > Not in the Layout test CI
> 
> Are LayoutTests run only on macOS then for every WebKit change?

I speak very poorly, sorry. The WPT test are executed as part of the Layout tests (step 9) and webdriver-test as well (step 19) [1]. The thing is they are encapsulated into the logic of the `Tools/Scripts/run-webkit-tests`.  

Also `run-webkit-tests` don't use a `wpt` from upstream but one imported into the WebKit  repo (I think with a small set of changes to work into the Layout-test suite).  the same happens with the w3c/wpt tests which are every then and now imported into the WebKit repo by running the `Tools/Scripts/import-w3c-tests` command. 

... and now, reading twice this, I know why you asking for this: ... _if some settings are enabled to make WPT pass on Linux for WebKit devs_ ...  -> Let me run it in local to extract the exact command line.


[1] https://build.webkit.org/builders/GTK%20Linux%2064-bit%20Release%20%28Tests%29/builds/11378

psaavedra commented:
@foolip Maybe you can help me with this I see many tests failing because the certificate issues.

![Screenshot from 2019-09-13 10-59-47](https://user-images.githubusercontent.com/376838/64850488-c5a39480-d615-11e9-978a-4519a412ed47.png)

To see those errors executed the tests exporting the DISPLAY to my X server hosted in the host OS of my laptop:

```
docker run -ti --rm \
       -e DISPLAY=$DISPLAY \
       -v /tmp/.X11-unix:/tmp/.X11-unix \
       5d794de51846
```
... and into de docker:

```
docker exec -it 9b29e24a11bc  /bin/bash
$ export DISPLAY=:0
./wpt run --channel=beta --metadata=infrastructure/metadata --log-wptreport=../artifacts/wpt_report.json --log-wptscreenshot=../artifacts/wpt_screenshot.txt --no-fail-on-unexpected --test-type=testharness --this-chunk=5 --total-chunks=15 --webdriver-binary=/usr/bin/WebKitWebDriver --webdriver-arg=--host=127.0.0.1 --log-tbpl-level=info --log-tbpl=- -y --no-pause --no-restart-on-unexpected --install-fonts --no-headless --verify-log-full epiphany
```

The result of those certificate issues is translated to TIMEOUT in the results. are tge CA certificate available in some place?

psaavedra commented:
I see ...

**NOTE**: The CA should not be installed in any browser profile used outside of tests, since it may be used to generate fake certificates. For browsers that use the OS certificate store, tests should therefore not be run manually outside a dedicated OS instance (e.g. a VM). To avoid this problem when running tests in Chrome or Firefox, use `wpt run`, which disables certificate checks and therefore doesn't require the root CA to be trusted.  (https://github.com/web-platform-tests/wpt/blob/master/tools/certs/README.md)

psaavedra commented:
> > Other searches worth looking at are `status:unknown` to find missing tests or subtests
> 
> OK. If I catched the idea it seems here I have bunch of errors related with webdriver to work on it: https://staging.wpt.fyi/results/webdriver/tests?diff&filter=ADC&q=status%3Aunknown&run_id=256970019&run_id=257120006

@foolip @clopez @Ms2ger --> I found the reason of these _TIMEOUTS_ ^ ... The execution of those test in epiphany is going slower than Safari. It takes more than 25 seconds to complete all the tests so the TestRunner interrupt the execution before the ending of the tasks.

I was able to pass those tests doing this hack in the code: https://github.com/psaavedra/wpt/commit/a7f3642cf9af7321211ac05cd06e0512fb4d6372

... but I still wonder if there is a more civilized way to setup different timeouts for a particular browser.

foolip commented:
It sounds like the most important thing is to make sure the certificate is installed somehow, either at the system level if webkitgtk uses that, or using some webkitgtk-specific mechanism.

Ignoring certificate errors sort of works, but isn't quite right. On at least Safari, the real certificate is used, installed at the OS level.

psaavedra commented:
> It sounds like the most important thing is to make sure the certificate is installed somehow, either at the system level if webkitgtk uses that, or using some webkitgtk-specific mechanism.

Yeah I did it: https://github.com/psaavedra/wpt/commit/3d1e95f1b85718ff949d9bf2f2a76cc7dddbe64b

psaavedra commented:
> > > Other searches worth looking at are `status:unknown` to find missing tests or subtests
> > 
> > 
> > OK. If I catched the idea it seems here I have bunch of errors related with webdriver to work on it: https://staging.wpt.fyi/results/webdriver/tests?diff&filter=ADC&q=status%3Aunknown&run_id=256970019&run_id=257120006
> 
> @foolip @clopez @Ms2ger --> I found the reason of these _TIMEOUTS_ ^ ... The execution of those tests in epiphany is going slower than in Safari. It takes more than 25 seconds to complete all the tests so the TestRunner interrupts the execution before the ending of the all tasks.
> 
> I was able to pass those tests doing this hack in the code: [psaavedra@a7f3642](https://github.com/psaavedra/wpt/commit/a7f3642cf9af7321211ac05cd06e0512fb4d6372)
> 
> ... but I still wonder if there is a more civilized way to setup different timeouts for a particular browser.

* Before the hack the timeout: https://staging.wpt.fyi/results/?diff&filter=ADC&q=status%3Aunknown&run_id=256970019&run_id=247310005
* After the hack: https://staging.wpt.fyi/results/?diff&filter=ADC&q=status%3Aunknown&run_id=256970019&run_id=249280006

These issues was affecting to many tests in `encoding`, `html` in  `webdriver`: 

![Screenshot from 2019-09-14 07-42-07](https://user-images.githubusercontent.com/376838/64904045-32756800-d6c4-11e9-82ba-56bfa586b93c.png)

_(Left: new tests increasing the timeout value / Right: previous results)_


There are still many other suites not yet working, for example: selection, server-timing, pointer-events, ...

foolip commented:
Regarding the timeouts, do you think webkitgtk is just generally slow and needs a longer timeout across the board, or is there a specific class of tests that's timing out more?

Note that many tests are flaky and might alter between pass and timeout. That could also be the cause of seeming regressions.

@lukebjerring with the tools we have today, what would be the best way to exclude flaky tests from an analysis?

psaavedra commented:
On Sat, Sep 14, 2019, 11:13 Philip Jägenstedt <notifications@github.com>
wrote:

> Regarding the timeouts, do you think webkitgtk is just generally slow and
> needs a longer timeout across the board, or is there a specific class of
> tests that's timing out more?
>

I need to study test by test not sure if it takes or not more time for
all the tests or just for a specific subset of them.

I think also that those times are quite sensible to the machine which
execute those tests.  Is there an easy way to compare the times for each
result and the machine which executes each one?

Note that many tests are flaky and might alter between pass and timeout.
> That could also be the cause of seeming regressions.
>

For the case of those that I debugged, related to the Webdriver, I don't
see those as flaky... just expending more time than the expected.

@lukebjerring <https://github.com/lukebjerring> with the tools we have
> today, what would be the best way to exclude flaky tests from an analysis?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/web-platform-tests/wpt/pull/18595?email_source=notifications&email_token=AAC4ABVGONQAZ76AUD6NCULQJSTK5A5CNFSM4IOIPPHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6WX6QQ#issuecomment-531464002>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC4ABSAYLXUSR7DA75XJODQJSTK5ANCNFSM4IOIPPHA>
> .
>

lukebjerring commented:
We don't yet have metadata on flakiness available; only queries that identify recently-flaky results.

foolip commented:
@psaavedra one thing you could try to reduce the impact of flaky tests is to trigger the exact same config twice for the two things you want to compare, giving you 4 runs. Then use `seq` on wpt.fyi to construct a query where the results for both configs are consistent and there's a regression.

https://web-platform-tests.org/running-tests/from-ci.html
https://github.com/web-platform-tests/wpt.fyi/blob/master/api/query/README.md#sequential

clopez commented:
Regarding the WebDriver related tests, I have discovered that several of the timeouts and failures are only reproducible with Epiphany but not with the WebKitGTK MiniBrowser.

I reported this on the webkit bugzilla at https://bugs.webkit.org/show_bug.cgi?id=202047 and I have been investigating it, but still I don't know what is wrong.

However, the issue with the timeout value also affects the MiniBrowser (but less)

Check for example this runs: (all done on the same docker container with the same WebKitGTK version):

1) With the @psaavedra hack for the timeout:

    * MiniBrowser:
	  ```
	  Ran 149 tests finished in 1148.6 seconds.
	    • 87 ran as expected. 0 tests skipped.
	    • 1 tests timed out unexpectedly
	    • 61 tests had unexpected subtest results
	  ```
	  https://staging.wpt.fyi/results/?run_id=237430004 (Note that it appears as epiphany because wpt.fyi doesn't seem to accept a product named "webkit" so I had to edit the json and wrote epiphany in product)


    * Epiphany
      ```
      Ran 149 tests finished in 856.5 seconds.
        • 85 ran as expected. 0 tests skipped.
        • 7 tests timed out unexpectedly
        • 57 tests had unexpected subtest results
      ```
      https://staging.wpt.fyi/results/?run_id=251400002


2) Without the @psaavedra timeout hack (WPT master)

    * MiniBrowser:
      ```
      Ran 151 tests finished in 390.3 seconds.
        • 65 ran as expected. 0 tests skipped.
        • 6 tests timed out unexpectedly
        • 80 tests had unexpected subtest results
      ```
      https://staging.wpt.fyi/results/?run_id=229570014 (Note that it appears as epiphany because wpt.fyi doesn't seem to accept a product named "webkit" so I had to edit the json and wrote epiphany in product)


    * Epiphany
      ```
      Ran 151 tests finished in 408.9 seconds.
        • 51 ran as expected. 0 tests skipped.
        • 17 tests timed out unexpectedly
        • 83 tests had unexpected subtest results
      ```
      https://staging.wpt.fyi/results/?run_id=195580002

foolip commented:
@clopez can you file a bug on wpt.fyi to accept the product name you want to use? Since there are many WebKit ports I wonder if it should be something other than just "webkit"?

clopez commented:
> @clopez can you file a bug on wpt.fyi to accept the product name you want to use? Since there are many WebKit ports I wonder if it should be something other than just "webkit"?

That's a very good point. I also start to find the name "webkit" kind of confusing. I have opened a new PR here to add a new product for this (webkitgtk_minibrowser). See https://github.com/web-platform-tests/wpt/pull/19227

I would open a request to add this product on wpt.fyi once accepted on wpt itself

foolip commented:
We'll need to add something in https://github.com/web-platform-tests/wpt.fyi/blob/25cb8670af7dbd802fb03c080bb1999f279c2031/results-processor/wptreport.py#L552-L554 to change the product name on wpt.fyi.

Is webkitgtk good as the main product name, and does that have a logo we can use?

clopez commented:
> Is webkitgtk good as the main product name, and does that have a logo we can use?

webkitgtk is good as main product name, but unfortunately it doesn't have a logo (AFAIK). 

I have opened PR https://github.com/web-platform-tests/wpt.fyi/pull/1509 adding it

clopez commented:
I have sent a run a complete run of task-cluster from [my test-taskcluster-webkitgtk branch](https://github.com/clopez/wpt/commits/test-taskcluster-webkitgtk) (forked from 19dc5b5b3c) for both [epiphany](https://staging.wpt.fyi/results/?run_id=237530004) and [webkitgtk_minibrowser](https://staging.wpt.fyi/results/?run_id=257490002) to wpt.fyi 

Here is the diff: https://staging.wpt.fyi/results/?diff&filter=ADC&run_id=237530004&run_id=257490002

A few highlights about the diffs:
 * `encoding/`: The 999 subtest-diff all seem to come from the test ```encoding/legacy-mb-korean/euc-kr/euckr-encode-form-cseuckr.html``` that gave timeout on epiphany, perhaps its a flaky test.
 * `webdriver/`: On Epiphany many webdriver tests give timeout. Tracked  in https://bugs.webkit.org/show_bug.cgi?id=202047
 * `css/`: There are several css tests that pass on Epiphany but fail on MiniBrowser in a very subtle way like [css/CSS2/bidi-text/bidi-box-model-003.xht](https://staging.wpt.fyi/analyzer?screenshot=sha1%3Ab70edc4e68cf8a1627471086fd96db50315b15e1&screenshot=sha1%3Af7d2e61c82d61655554d610e8db48a1a37a6fd9f). Perhaps its related to the hardware acceleration (wheter to use OpenGL or not) policy that is different on Epiphany (disabled) and MiniBrowser (on-demand).

clopez commented:
@foolip I have evaluated doing builds of webkitgtk for Ubuntu 18.04 instead vs Debian testing/unstable and I think now that it will be better to continue with Debian. The reason is that some of dependencies on Ubuntu 18.04 are not available or too much old. To name them:

* For testing MSE support with trunk now you need gstreamer-1.16 (Ubuntu 18.04 has 1.14). 
* To enable the new default render path you need libwpe, which is still not available on Ubuntu 18.04
* The version of GTK (3.22) is enough for WebKitGTK, but not for Epiphany. If we want to run Epiphany tests (with last version of Epiphany) that would make things really hard.

foolip commented:
That makes sense, let's stick with Debian! Is there any additional cleanup of this PR needed, or should I review it in detail now?

clopez commented:
> That makes sense, let's stick with Debian! Is there any additional cleanup of this PR needed, or should I review it in detail now?

This still needs several cleanups. I will update the patch when its ready for review.

clopez committed:
Add Dockerfile for WebKitGTK testing based on Debian. (#18595).

* This adds a new Dockerfile that will be used for tests
with webkitgtk_minibrowser on taskcluster.

* The image is configured to install the WebKitGTK packages
from a repository at https://webkitgtk.org/debian
This repository will provide updated packages of WebKitGTK
releases with experimental features enabled.
(some of this features need to be enabled at build-time,
therefore the need for custom builds)

* The image is based on Debian 10 (Buster) instead of Ubuntu,
because the current release bot at build.webkit.org [1]
runs Debian 10. And having this image match what this bot
runs, will allow binary compatibility with the built products
this bot is producing. The idea is (in a next patch) try
to download and use those products to run WPT tests with
the nightly builds from webkitgtk trunk/ToT.

[1] https://build.webkit.org/builders/GTK%20Linux%2064-bit%20Release%20%28Build%29

clopez commented:
In the end I have decided to go with Debian stable (10/buster) because I think it will allow us to test with webkitgtk trunk/master/ToT for free (without having to build it)

My idea is to re-use [the official built products that the WebKitGTK release bot at build.webkit.org is producing](https://build.webkit.org/builders/GTK%20Linux%2064-bit%20Release%20%28Build%29). This bot is running Debian stable, so by having this docker image run the same distro/version I believe binary compatibility will be granted. I just need to find some time to create a script to automate packing the built-product this bot produces and make it available for download. That way we could run the tests using trunk/ToT without having to build it, just download this product.

And for the tests with official releases (tarballs) I have created a debian repository at https://webkitgtk.org/debian/ with builds from this tarballs, but with experimental features enabled. Some of this features (ex: webrtc or service workers) are needed to enable at build-time, therefore this need for custom builds. So, this docker image will be configured to pull all the webkitgtk related packages from that repository.

Regarding the repository, generating new packages is itself [automated by another bot](https://build-webkit.igalia.com/builders/GTK%20Linux%2064-bit%20Debian%20Package%20%28WPT%20Tests%29) that pulls new changes from [a fork we did of the official debian webkitgtk packages](https://gitlab.com/webkit-packaging-wpt/webkit-debian). We intend to keep this repository updated with builds with experimental features enabled for each new webkitgtk release.

clopez commented:
@foolip I have updated the branch of this PR. If you can review it that will be great.
In order to start running tests on taskcluster some changes to .taskcluster.yml and to the webkitgtk_minibrowser driver are needed on top of this. I will push those changes tomorrow into a new pull-request. This one is only for the new dockerfile.

foolip commented on tools/docker/README.md:
````diff
@@ -0,0 +1,19 @@
+docker push <tag>
+```
+
+(This image is not yet used in .taskcluster.yml.)
````
Remember to remove this line in the next PR :)
